Fix and migration of textures & models for 8.2 API changes

mrbuds opened the pull request:
# Description
Fixes #1291
SetTexture() can't use internal Path like "Texture/...." and need an id instead.
I replaced path for some textures with what the function GetFileIDFromPath() return on live, and fixed texture browser to handle numbers.

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
- [x] Test on live
- [x] Test on 8.2 ptr

# TODO
- [x] Test everything on PTR (currently down)

Model browser is still broken, all path in the file "ModelPaths.lua" has to be replaced by ids.
Afaik all ids can be found in https://raw.githubusercontent.com/Resike/BlizzardInterfaceResources/b6abf3b3b320a2a2f1054f20ddd17df6b1a0404c/Resources/Data/CreatureDisplayPath.lua
But i'd like @InfusOnWoW on the matter first because you are the last one who edited the file, from a list submited on reddit from what you wrote on commit message, and the whole list might needs an update.

emptyrivers commented:
7.2 PTR was over a year ago, so I think this is a bit late ;)

mrbuds commented:
shit too late :(

Stanzilla commented:
So for all the upcoming changes for file name to id stuff, I want the original file name as a comment next to it, just so it is clear what it is from just reading the code

Stanzilla commented on WeakAuras/Types.lua:
```diff
@@ -880,15 +879,15 @@ WeakAuras.texture_types = {
     ["bonusobjectives-bar-spark"] = "Bonus Objectives Spark",
-    ["worldstate-capturebar-spark-green"] = "Capture Bar Green Spark",
-    ["worldstate-capturebar-spark-yellow"] = "Capture Bar Yellow Spark"
+    -- ["worldstate-capturebar-spark-green"] = "Capture Bar Green Spark", -- removed in 7.2
```
GetBuildInfo() check if you want

InfusOnWoW commented:
I only update the ModelPaths.lua file if someone asks for a update, so it's almost always out of date. The commit message does link to a small C++ utility I wrote to generate the file, and I can adjust that to read in the file that you posted, but:

* That list does not include the spells directory
* And what should we do in the future, when we no longer get file names?

mrbuds commented:
added a fix for texture/progressTexture input widget not working with numbers
added a migrations for textures

models migrations & browser are still TODO

mrbuds commented:
Added a temporary file that will replace ModelPaths.lua after 8.2 is out
Made with the following addon: [WAModelIds.zip](https://github.com/WeakAuras/WeakAuras2/files/3110110/WAModelIds.zip)

mrbuds commented on WeakAuras/WeakAuras.lua:
```diff
@@ -2868,6 +2872,38 @@ function WeakAuras.Modernize(data)
+    end
+  end
+
+  if data.regionType == "model" and WeakAuras.BuildInfo <= 80100 and not data.model_fileId then -- prepare for migration at 8.2
```
model_fileId cannot be automatically populated after 8.2

mrbuds commented on WeakAuras/WeakAuras.lua:
```diff
@@ -5572,3 +5608,21 @@ function WeakAuras.FindUnusedId(prefix)
+    if isUnit then
+      pcall(function() frame:SetUnit(data) end)
+    else
+      pcall(function() frame:SetModel(data) end)
```
all model setters are wrapped with pcall to not show lua error on bad inputs

mrbuds commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+	line.parent = parent
+	line.visible = v.visible
+	line.uniquevalue = GetButtonUniqueValue(line)
+	line.fileId = v.fileId
```
read new fileId property added to ModelPaths.lua

mrbuds commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+		local status = self.status or self.localstatus
+		if status.selected ~= value then
+			status.selected = value
+			self:Fire("OnGroupSelected", value, fileId)
```
return value and fileId on "OnGroupSelected" event

mrbuds commented on WeakAurasOptions/AceGUI-Widgets/AceGUIWidget-WeakAurasTextureButton.lua:
```diff
@@ -112,7 +112,7 @@ local function Constructor()
   texture:SetPoint("TOPRIGHT", button, -4, -4);
 
-  button:SetScript("OnEnter", function() Show_Tooltip(button, texture.name, texture.path:gsub("\\", "\n")) end);
+  button:SetScript("OnEnter", function() Show_Tooltip(button, texture.name, type(texture.path) == "string" and texture.path:gsub("\\", "\n") or texture.path) end);
```
i added a string type check because previously i migrated texture.path to a number, but reverted it, and i still had some data with type == "number"
this can be safely reverted

Stanzilla replied:
revert it!

mrbuds commented on WeakAurasOptions/OptionsFrames/TexturePicker.lua:
```diff
@@ -159,8 +159,10 @@ local function ConstructTexturePicker(frame)
-    if(type(self.givenPath) == "string") then
+    if type(self.givenPath) == "string" then
       givenPath = self.givenPath;
+    elseif type(self.givenPath) == "number" then
```
same here can be reverted

Stanzilla replied:
do it :)

mrbuds commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -188,6 +182,22 @@ local function createOptions(id, data)
     },
   };
 
+  if WeakAuras.BuildInfo > 80100 then
```
region option switch to model_fileId after 8.2

InfusOnWoW commented on WeakAuras/WeakAuras.lua:
```diff
@@ -2868,6 +2872,38 @@ function WeakAuras.Modernize(data)
+    local modelId = modelMigration:GetModelFileID()
+    if modelId then
+      data.model_fileId = tostring(modelId)
+    end
```
```suggestion
    elseif data.modelIsUnit then
      data.model_fileId = data.model_path
    end
```

mrbuds commented:
Migration of models that were using SetDisplayInfo should work fine, WeakAuras.SetModel works as intended with numbers pre 8.2
Your example with old id "136" get converted to model_fileId "122968" which is creature/arthaslichking/arthaslichking.m2 according to https://wow.tools/files/#search=122968&page=1&sort=0&desc=asc
But for a reason i don't know SetDisplayInfo(136) doesn't show anything, i tested with master branch and get the same result, so it's not a new issue from my code.

InfusOnWoW commented:
I can't test this on the PTR, as it is down atm. I think it should now work, but testing it on the PTR would be good.

mrbuds commented:
I thought from tests i did that SetDisplay wasn't working anymore in 8.2 but i was wrong, and indeed it doesn't do the exact same thing, good catch.

About the layout i think "animate" toggle should stay on same line as it's slider, but i'm not sure about moving "Portrait Zoom" on the same line as the "SetTransform" toggle.

![img](https://i.imgur.com/cnAoY8G.png)

What do you think ?

Stanzilla commented:
Try it and post before/after screenshots! :P

mrbuds commented:
before
![img](https://i.imgur.com/01hUmue.png)

after
![img](https://i.imgur.com/jYiA22T.png)

Stanzilla commented:
after looks better to me

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+WeakAurasTreeGroup Container
+Container that uses a tree control to switch between groups.
+-------------------------------------------------------------------------------]]
+local Type, Version = "WeakAurasTreeGroup", 41
```
start with 1 instead of 41 maybe? :D

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+		local maxlines = (floor(((self.treeframe:GetHeight()or 0) - 20 ) / 18))
+		if maxlines <= 0 then return end
+
+		-- workaround for lag spikes on WoW 8.0
```
```suggestion
		-- WORKAROUND for lag spikes on WoW 8.0
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+				end
+				if not show then
+					-- selection was deleted or something?
+				elseif show>=first and show<=last then
```
```suggestion
				elseif show >= first and show <= last then
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+					-- all good
+				else
+					-- scrolling needed!
+					if show<first then
```
```suggestion
					if show < first then
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+				else
+					-- scrolling needed!
+					if show<first then
+						status.scrollvalue = show-1
```
```suggestion
						status.scrollvalue = show - 1
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+					else
+						status.scrollvalue = show-maxlines
+					end
+					first, last = status.scrollvalue+1, status.scrollvalue + maxlines
```
```suggestion
					first, last = status.scrollvalue + 1, status.scrollvalue + maxlines
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+
+				buttons[buttonnum] = button
+				button:SetParent(treeframe)
+				button:SetFrameLevel(treeframe:GetFrameLevel()+1)
```
```suggestion
				button:SetFrameLevel(treeframe:GetFrameLevel() + 1)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+						button:SetPoint("TOPLEFT", 0, -10)
+					end
+				else
+					button:SetPoint("TOPRIGHT", buttons[buttonnum-1], "BOTTOMRIGHT",0,0)
```
```suggestion
					button:SetPoint("TOPRIGHT", buttons[buttonnum - 1], "BOTTOMRIGHT", 0, 0)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+					end
+				else
+					button:SetPoint("TOPRIGHT", buttons[buttonnum-1], "BOTTOMRIGHT",0,0)
+					button:SetPoint("TOPLEFT", buttons[buttonnum-1], "BOTTOMLEFT",0,0)
```
```suggestion
					button:SetPoint("TOPLEFT", buttons[buttonnum - 1], "BOTTOMLEFT", 0, 0)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+				end
+			end
+
+			UpdateButton(button, line, status.selected == line.uniquevalue, line.hasChildren, groupstatus[line.uniquevalue] )
```
```suggestion
			UpdateButton(button, line, status.selected == line.uniquevalue, line.hasChildren, groupstatus[line.uniquevalue])
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+		if show then
+			self.scrollbar:Show()
+			if self.buttons[1] then
+				self.buttons[1]:SetPoint("TOPRIGHT", self.treeframe,"TOPRIGHT",-22,-10)
```
```suggestion
				self.buttons[1]:SetPoint("TOPRIGHT", self.treeframe, "TOPRIGHT", -22, -10)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+		else
+			self.scrollbar:Hide()
+			if self.buttons[1] then
+				self.buttons[1]:SetPoint("TOPRIGHT", self.treeframe,"TOPRIGHT",0,-10)
```
```suggestion
				self.buttons[1]:SetPoint("TOPRIGHT", self.treeframe, "TOPRIGHT", 0, -10)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+	scrollbar:SetScript("OnValueChanged", nil)
+	scrollbar:SetPoint("TOPRIGHT", -10, -26)
+	scrollbar:SetPoint("BOTTOMRIGHT", -10, 26)
+	scrollbar:SetMinMaxValues(0,0)
```
```suggestion
	scrollbar:SetMinMaxValues(0, 0)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+
+	local scrollbg = scrollbar:CreateTexture(nil, "BACKGROUND")
+	scrollbg:SetAllPoints(scrollbar)
+	scrollbg:SetColorTexture(0,0,0,0.4)
```
```suggestion
	scrollbg:SetColorTexture(0, 0, 0, 0.4)
```

Stanzilla commented on WeakAurasOptions/AceGUI-Widgets/AceGUIContainer-WeakAurasTreeGroup.lua:
```diff
@@ -0,0 +1,719 @@
+	scrollbg:SetAllPoints(scrollbar)
+	scrollbg:SetColorTexture(0,0,0,0.4)
+
+	local border = CreateFrame("Frame",nil,frame)
```
```suggestion
	local border = CreateFrame("Frame", nil, frame)
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -188,6 +186,30 @@ local function createOptions(id, data)
+      order = 0.6,
+      hidden = function() return data.modelIsUnit end
+    }
+
```
remove space

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -265,6 +270,7 @@ end
 local function createIcon()
   local data = {
     model_path = "Creature/Arthaslichking/arthaslichking.m2",
+    model_fileId = "122968",
```
```suggestion
    model_fileId = "122968", -- Creature/Arthaslichking/arthaslichking.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -296,6 +302,7 @@ local templates = {
       width = 100,
       height = 100,
       model_path = "spells/6fx_smallfire.m2",
+      model_fileId = "937416",
```
```suggestion
      model_fileId = "937416", -- spells/6fx_smallfire.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -310,6 +317,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_druid_halfmoon_missile.m2",
+      model_fileId = "1322288",
```
```suggestion
      model_fileId = "1322288", -- spells/7fx_druid_halfmoon_missile.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -324,6 +332,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/proc_arcane_impact_low.m2",
+      model_fileId = "1042743",
```
```suggestion
      model_fileId = "1042743", -- spells/proc_arcane_impact_low.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -338,6 +347,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_godking_orangerune_state.m2",
+      model_fileId = "1307356",
```
```suggestion
      model_fileId = "1307356", -- spells/7fx_godking_orangerune_state.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -349,6 +359,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_godking_bluerune_state.m2",
+      model_fileId = "1307354",
```
```suggestion
      model_fileId = "1307354", -- spells/7fx_godking_bluerune_state.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -360,6 +371,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_godking_yellowrune_state.m2",
+      model_fileId = "1307358",
```
```suggestion
      model_fileId = "1307358", -- spells/7fx_godking_yellowrune_state.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -371,6 +383,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_godking_purplerune_state.m2",
+      model_fileId = "1307355",
```
```suggestion
      model_fileId = "1307355", -- spells/7fx_godking_purplerune_state.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Model.lua:
```diff
@@ -382,6 +395,7 @@ local templates = {
       advance = true,
       sequence = 1,
       model_path = "spells/7fx_godking_greenrune_state.m2",
+      model_fileId = "1307357",
```
```suggestion
      model_fileId = "1307357", -- spells/7fx_godking_greenrune_state.m2
```

Stanzilla commented on WeakAurasOptions/RegionOptions/ProgressTexture.lua:
```diff
@@ -764,7 +764,7 @@ local templates = {
       yOffset = 150,
       mirror = true,
-      foregroundTexture = "Textures\\SpellActivationOverlays\\Backlash",
+      foregroundTexture = "460830",
```
```suggestion
      foregroundTexture = "460830", -- "Textures\\SpellActivationOverlays\\Backlash"
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Texture.lua:
```diff
@@ -182,7 +182,7 @@ local templates = {
     title = L["Star"],
     data = {
-      texture = "Spells\\T_Star3",
+      texture = "241049",
```
```suggestion
      texture = "241049", -- Spells\\T_Star3
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Texture.lua:
```diff
@@ -192,7 +192,7 @@ local templates = {
     title = L["Leaf"],
     data = {
-      texture = "Spells\\Nature_Rune_128",
+      texture = "166606",
```
```suggestion
      texture = "166606", -- Spells\\Nature_Rune_128
```

Stanzilla commented on WeakAurasOptions/RegionOptions/Texture.lua:
```diff
@@ -202,7 +202,7 @@ local templates = {
     title = L["Hawk"],
     data = {
-      texture = "Spells\\Aspect_Hawk",
+      texture = "165609",
```
```suggestion
      texture = "165609", -- Spells\\Aspect_Hawk
```

Stanzilla commented on WeakAurasOptions/WeakAurasOptions.toc:
```diff
@@ -64,3 +64,4 @@ AceGUI-Widgets\AceGUIWidget-WeakAurasNewButton.lua
 AceGUI-Widgets\AceGUIWidget-WeakAurasImportButton.lua
 AceGUI-Widgets\AceGUIWidget-WeakAurasSortedDropDown.lua
 AceGUI-Widgets\AceGUIWidget-WeakAurasTwoColumnDropDown.lua
+AceGUI-Widgets\AceGUIContainer-WeakAurasTreeGroup.lua
```
we currently don't use localization that file but add to `babelfish.lua` anyway
